fix: add network interface binding support

roderickvd opened the pull request:
## Description

Default is IPv4-only to prevent AAAA DNS lookup timeouts since Deezer services are primarily IPv4. This fixes connection timeouts in dual-stack environments where failed IPv6 attempts would delay IPv4 connectivity.

## Related Issues

- Fixes #52

## Implementation Details

Add ability to bind outgoing connections to specific network interfaces:
- New `--bind` CLI option (defaults to 0.0.0.0)
- Configurable through `PLEEZER_BIND` env var
- Supports both IPv4 and IPv6 addresses
- Useful for VPN/tunnel routing or multi-homed systems

## Testing

Tested locally, but do not have access to dual-home environment.

## Due Diligence

Please confirm that you have completed the following tasks by checking the boxes:

- [x] I have linked any related [issues or feature requests](https://github.com/roderickvd/pleezer/issues).
- [x] I have selected the appropriate labels for this pull request.
- [x] I have performed cross-platform testing if possible.
- [x] I have updated the [CHANGELOG.md](https://github.com/roderickvd/pleezer/blob/main/CHANGELOG.md) file with a summary of my changes under the "Unreleased" section.
- [x] I have kept the pull request as a draft until it is ready for review (if applicable).
- [x] I have read and understood the [Contributing guidelines](https://github.com/roderickvd/pleezer/blob/main/CONTRIBUTING.md).

roderickvd committed:
fix: add network interface binding support

Add ability to bind outgoing connections to specific network interfaces:
- New --bind CLI option (defaults to 0.0.0.0)
- Configurable through PLEEZER_BIND env var
- Supports both IPv4 and IPv6 addresses
- Useful for VPN/tunnel routing or multi-homed systems

Default is IPv4-only to prevent AAAA DNS lookup timeouts since Deezer
services are primarily IPv4. This fixes connection timeouts in dual-stack
environments where failed IPv6 attempts would delay IPv4 connectivity.

Fixes #52

scred47 commented:
Hi,

Unfortunately this hasn't help fix the timeout problem

```
isidro@pleezer:~ $ pleezer
[2025-01-30T14:34:19Z INFO  pleezer] starting pleezer/0.11.1.79dc0f9 (2025-01-29); release
[2025-01-30T14:34:19Z INFO  pleezer] parsing secrets from secrets.toml
[2025-01-30T14:34:24Z ERROR pleezer] operation timed out: error sending request for url (https://www.deezer.com/en/channels/explore/)
```


With aaaa lookups not disabled it times out ans still sends an AAAA lookup  

```
isidro@pleezer:~ $ sudo tcpdump -i wlan0 port 53
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on wlan0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
15:34:19.402870 IP 192.168.2.187.43424 > pi.hole.domain: 17247+ A? www.deezer.com. (32)
15:34:19.402915 IP 192.168.2.187.43424 > pi.hole.domain: 3692+ AAAA? www.deezer.com. (32)
15:34:19.417832 IP 192.168.2.187.48667 > pi.hole.domain: 44016+ PTR? 187.2.168.192.in-addr.arpa. (44)
15:34:19.428544 IP pi.hole.domain > 192.168.2.187.48667: 44016 NXDomain* 0/0/0 (44)
15:34:19.435894 IP pi.hole.domain > 192.168.2.187.43424: 17247 6/0/0 CNAME web.geo2.deezer.com., CNAME adn-europe.deezer.com., CNAME www.deezer.com.edgekey.net., CNAME e28200.a.akamaiedge.net., A 2.16.27.80, A 2.16.27.99 (186)
15:34:19.436736 IP 192.168.2.187.52398 > pi.hole.domain: 30733+ PTR? 3.3.168.192.in-addr.arpa. (42)
15:34:19.450784 IP pi.hole.domain > 192.168.2.187.52398: 30733* 1/0/0 PTR pi.hole. (63)
```
as mentioned in https://github.com/roderickvd/pleezer/issues/52 I would lean towards this just being the OS behaving in an unexpected way. sending AAAA lookups makes no sense in a IPv4 only  network  with no Ipv6 configured on the host and is creating problems not restricted to pleezer. No mention also of the unnecessary noise.

scred47 commented:
Hi,

Unfortunately this hasn't help fix the timeout problem

```
isidro@pleezer:~ $ pleezer
[2025-01-30T14:34:19Z INFO  pleezer] starting pleezer/0.11.1.79dc0f9 (2025-01-29); release
[2025-01-30T14:34:19Z INFO  pleezer] parsing secrets from secrets.toml
[2025-01-30T14:34:24Z ERROR pleezer] operation timed out: error sending request for url (https://www.deezer.com/en/channels/explore/)
```


With aaaa lookups not disabled it times out ans still sends an AAAA lookup  

```
isidro@pleezer:~ $ sudo tcpdump -i wlan0 port 53
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on wlan0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
15:34:19.402870 IP 192.168.2.187.43424 > pi.hole.domain: 17247+ A? www.deezer.com. (32)
15:34:19.402915 IP 192.168.2.187.43424 > pi.hole.domain: 3692+ AAAA? www.deezer.com. (32)
15:34:19.417832 IP 192.168.2.187.48667 > pi.hole.domain: 44016+ PTR? 187.2.168.192.in-addr.arpa. (44)
15:34:19.428544 IP pi.hole.domain > 192.168.2.187.48667: 44016 NXDomain* 0/0/0 (44)
15:34:19.435894 IP pi.hole.domain > 192.168.2.187.43424: 17247 6/0/0 CNAME web.geo2.deezer.com., CNAME adn-europe.deezer.com., CNAME www.deezer.com.edgekey.net., CNAME e28200.a.akamaiedge.net., A 2.16.27.80, A 2.16.27.99 (186)
15:34:19.436736 IP 192.168.2.187.52398 > pi.hole.domain: 30733+ PTR? 3.3.168.192.in-addr.arpa. (42)
15:34:19.450784 IP pi.hole.domain > 192.168.2.187.52398: 30733* 1/0/0 PTR pi.hole. (63)
```
as mentioned in https://github.com/roderickvd/pleezer/issues/52 I would lean towards this just being the OS behaving in an unexpected way. sending AAAA lookups makes no sense in a IPv4 only  network  with no Ipv6 configured on the host and is creating problems not restricted to pleezer. No mention also of the unnecessary noise.

roderickvd commented:
Thanks a lot for testing. I feared as much. I will keep the `bind` option as nice to have. 

Maybe best to set the timeouts to just over 5 seconds then.

Implementing a custom DNS resolver would require making choices about DNS resolvers outside of the system configuration. That’s getting into an area of user privacy and foot-shooting (“why is pleezer not using my name servers?”) that I don’t want to go.

roderickvd committed:
refactor: rename secrets file parameter for consistency

Rename secrets file parameter to just `secrets` across the codebase for
better consistency with other parameter names.

Also document the new `bind` parameter in the README.

roderickvd committed:
fix: adjust timeouts for better DNS resolution

- Increase connection timeout to 10s to accommodate AAAA record timeouts
  that can occur on Linux systems
- Reduce read timeout to 2s as individual reads don't need the full 5s

This helps prevent connection failures in dual-stack environments while
keeping individual reads responsive.

roderickvd commented:
I just pushed a new attempt with the read and connect timeouts tuned to respectively 2 and 10 seconds.
If you would give it a whirl?

roderickvd committed:
refactor: rename bind to bind_address for clarity

Rename 'bind' to 'bind_address' across the codebase to better indicate
its purpose as an IP address configuration. This makes the naming more
explicit and consistent with its usage.

The change affects internal field names only, the CLI option remains
unchanged.

roderickvd commented:
I am gonna go ahead and merge this, because I'd like to get a release out of the door. If we need to further tune the timeout, please reopen the issue and it can go into a patch release.

scred47 commented:
Hi,

> Implementing a custom DNS resolver would require making choices about DNS resolvers outside of the system configuration. That’s getting into an area of user privacy and foot-shooting (“why is pleezer not using my name servers?”) that I don’t want to go.

Fair enough. 

>I just pushed a new attempt with the read and connect timeouts tuned to respectively 2 and 10 seconds.
If you would give it a whirl?

Sure thing. Apologies i couldn' test earlier. The new behavior is that it times out after 2 seconds.

```
:~ $ pleezer
[2025-02-01T09:09:58Z INFO  pleezer] starting pleezer/0.12.0.529af50 (2025-01-31); release
[2025-02-01T09:09:58Z INFO  pleezer] parsing secrets from secrets.toml
[2025-02-01T09:10:00Z ERROR pleezer] operation timed out: error sending request for url (https://www.deezer.com/en/channels/explore/)
```

My opinion is that we shouldn' t mess around with the http timeouts because they work fine if DNS works fine, and maybe just document the possible underlying problem and the workarounds in the troubleshooting section of Readme. what is your take on that?

I found a second workaround which is to just change the default dns timeout to 1 second. This way if AAAA lookup are slow for reasons that are beyond user's control, the process doesn' t wait for the AAAA lookup reply.

```
:~ $ cat /etc/resolv.conf
options timeout:1
nameserver 195.121.1.66
```

The result is that with the same version that it connects fine.

```
:~ $ pleezer
[2025-02-01T09:34:55Z INFO  pleezer] starting pleezer/0.12.0.529af50 (2025-01-31); release
[2025-02-01T09:34:55Z INFO  pleezer] parsing secrets from secrets.toml
[2025-02-01T09:34:57Z INFO  pleezer::remote] logging in with email and password
[2025-02-01T09:34:57Z INFO  pleezer::remote] ready for discovery

```

tcpdump shows the connection is initiated 1 second after the A record lookup.

```

10:34:55.289692 IP 192.168.2.187.46199 > 195.121.1.66.53: 57595+ A? www.deezer.com. (32)
10:34:55.289900 IP 192.168.2.187.46199 > 195.121.1.66.53: 16623+ AAAA? www.deezer.com. (32)
10:34:55.303810 IP 195.121.1.66.53 > 192.168.2.187.46199: 57595 6/0/0 CNAME web.geo2.deezer.com., CNAME adn-europe.deezer.com., CNAME www.deezer.com.edgekey.net., CNAME e28200.a.akamaiedge.net., A 2.16.6.21, A 2.16.6.9 (186)
10:34:56.290198 IP 192.168.2.187.46199 > 195.121.1.66.53: 57595+ A? www.deezer.com. (32)
10:34:56.305812 IP 195.121.1.66.53 > 192.168.2.187.46199: 57595 6/0/0 CNAME web.geo2.deezer.com., CNAME adn-europe.deezer.com., CNAME www.deezer.com.edgekey.net., CNAME e28200.a.akamaiedge.net., A 2.16.6.21, A 2.16.6.9 (186)
10:34:56.305999 IP 192.168.2.187.46199 > 195.121.1.66.53: 16623+ AAAA? www.deezer.com. (32)
10:34:56.320922 IP 195.121.1.66.53 > 192.168.2.187.46199: 16623 4/1/0 CNAME web.geo2.deezer.com., CNAME adn-europe.deezer.com., CNAME www.deezer.com.edgekey.net., CNAME e28200.a.akamaiedge.net. (212)
10:34:56.321902 IP 192.168.2.187.42767 > 2.16.6.21.443: Flags [S], seq 1681036864, win 64240, options [mss 1460,sackOK,TS val 498926662 ecr 0,nop,wscale 6], length 0
```

roderickvd commented:
> > I just pushed a new attempt with the read and connect timeouts tuned to respectively 2 and 10 seconds.
> > If you would give it a whirl?
> 
> Sure thing. Apologies i couldn' test earlier. The new behavior is that it times out after 2 seconds.

No problem. 5a9d981c867acd821face6991ad398656c3c7cb5 tunes the parameters again.

I thought the `CONNECT_TIMEOUT` would deal with the DNS resolution, but I was wrong: it deals with the TCP connection _after_ the DNS resolution. So that's back to 2s again and now `READ_TIMEOUT` is at 10s.

> My opinion is that we shouldn' t mess around with the http timeouts because they work fine if DNS works fine, and maybe just document the possible underlying problem and the workarounds in the troubleshooting section of Readme. what is your take on that?

I agree once Raspberry Pi OS gets this down. Until then, like you pointed out in #52, we can expect RPi OS to be so common that I'd like to ensure an out-of-the-box experience. I've already found that users rather post issues than read READMEs, so it's for them _and_ me 😄

scred47 commented:
Hi,

No luck with that last build. It times out 2 seconds after.


```
:~ $ cat pleezer/src/http.rs | grep -e CONNECT_TIMEOUT -e READ_TIMEOUT
    const CONNECT_TIMEOUT: Duration = Duration::from_secs(2);
    const READ_TIMEOUT: Duration = Duration::from_secs(10);
            .connect_timeout(Self::CONNECT_TIMEOUT)
            .read_timeout(Self::READ_TIMEOUT)
:~ $ pleezer
[2025-02-04T07:55:59Z INFO  pleezer] starting pleezer/0.12.0.5a9d981 (2025-02-02); release
[2025-02-04T07:55:59Z INFO  pleezer] parsing secrets from secrets.toml
[2025-02-04T07:56:01Z ERROR pleezer] service unavailable: error sending request for url (https://www.deezer.com/en/channels/explore/)

```
No TCP connection is initiated in that process.

```
:~ $ sudo tcpdump -ni wlan0 port 53 or 80 or 443
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on wlan0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
08:55:59.556751 IP 192.168.2.187.43780 > 192.168.3.3.53: 49083+ A? www.deezer.com. (32)
08:55:59.556955 IP 192.168.2.187.43780 > 192.168.3.3.53: 17352+ AAAA? www.deezer.com. (32)
08:55:59.583268 IP 192.168.3.3.53 > 192.168.2.187.43780: 49083 6/0/0 CNAME web.geo2.deezer.com., CNAME adn-europe.deezer.com., CNAME www.deezer.com.edgekey.net., CNAME e28200.a.akamaiedge.net., A 2.16.6.9, A 2.16.6.21 (186)

```
so it seems that whichever is met first be it READ or CONNECT, it stops the process.

As a complement I built that on a raspberry 2B that was running Debian bullseye ( debian 11) and there is no issue with that one Same subnet, same DHCP server same DNS server) so next I will be upgrading that same device to debian 12 to see if the issue appears after upgrade.

roderickvd commented:
Dang. Instead of me trying to tune what I cannot reproduce, could you please let me know what the minimum working parameters are?

scred47 commented:
So the issue seems very specific to using the RPI zero 2W builtin wifi. It doesn't happen even on Debian 12 with RPI 2B using the builtin ethernet.

> Dang. Instead of me trying to tune what I cannot reproduce, could you please let me know what the minimum working parameters are?

sure. I'll play around with those numbers but I have to say that when we put very high timeouts , every time i hit next song The next song would skip to the song after that and a log saying "deadline exceeded" or "deadline reached" was shown.

It may take a while before I can test. I will try with 6 and 6 first. but i suspect this will break other stuff as everytime we are going to need to resolve the name to set an HTTP request we are going to go over the 5 seconds deadline.

roderickvd commented:
Huh, how about that. Yes if it's only one legacy platform in a specific environment, then I'm with you: I'd rather revert and point those few users to modifying their `/etc/resolv.conf`.
